### PR TITLE
feat(ingestion/powerbi): ingest powerbi app

### PR DESCRIPTION
**Pull request by sid-acryl**

### CODE REVIEW

**Comment by aviv-julienjehannet** [comment]
Strange error not strictly linked to my PR, I suppose:
https://github.com/datahub-project/datahub/actions/runs/11350486344/job/31568984855?pr=11629#step:12:1720

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py [hunk 1]
```diff
@@ -97,6 +99,8 @@ def __init__(
             ),
         )
 
+        self.get_app = lru_cache(maxsize=128)(self.__get_app)
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py:102]: why would we be calling get_app multiple times?

**Comment by sid-acryl** [comment]
> Strange error not strictly linked to my PR, I suppose: https://github.com/datahub-project/datahub/actions/runs/11350486344/job/31568984855?pr=11629#step:12:1720

It will be fixed now

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py (hunk 1 above)
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py]: Not needed, I removed the cache.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py [hunk 2]
```diff
@@ -221,6 +229,7 @@ def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
                 tiles=[],
                 users=[],
                 tags=[],
+                app_reference=None,  # It is getting set later from scan_result
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py:232]: what do we use app_reference for?

**Comment by Daniellundin048** [comment]
respond

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py [hunk 3]
```diff
@@ -1306,6 +1311,102 @@ def extract_independent_datasets(
...
+    ) -> Iterable[MetadataWorkUnit]:
+        if workspace.app is None:
+            return
+
+        if not self.source_config.extract_app:
+            self.reporter.info(
+                title="App Ingestion Is Disabled",
+                message="You are missing workspace app metadata. Please set flag `extract_app` to `true` in recipe to ingest workspace app.",
+                context=f"workspace-name={workspace.name}, app-name = {workspace.app.name}",
+            )
```
[the remark this answers is not in the document]
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:1325]: Added return. .app will be populated on next run after setting `extract_app` to true

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py [hunk 4]
```diff
@@ -680,6 +721,15 @@ def profile_dataset(
 
         table.column_count = column_count
 
+    def _get_app(
+        self,
+        app_id: str,
+    ) -> Optional[Dict]:
+        # [Date: 2024/10/18] As per API doc, the service principal approach is not supported for regular API
+        # https://learn.microsoft.com/en-us/rest/api/power-bi/apps/get-app
```
[the remark this answers is not in the document]
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py:729]: It is not going to impact user experience, as we are getting app from AdminApiResolver and I have updated the assets list in doc MD file.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py [hunk 5]
```diff
@@ -423,6 +427,83 @@ def _get_workspace_datasets(self, workspace: Workspace) -> dict:
...
+        if app is None:
+            self.__reporter.info(
+                title="App Not Found",
+                message="The workspace includes an app, but its metadata is missing from the API response.",
+                context=f"workspace_name={workspace.name}",
+            )
+            return
+
+        # Map to find out which dashboards belongs to the App
+        dashboard_map: Dict[str, Dict] = {
```
[the remark this answers is not in the document]
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py:485]: done

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py (hunk 2 above)
Comment (sid-acryl) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py]: Removed it, not in used

**Review by hsheth2 — APPROVED** [review]

**Pull request merged**